File: src/lib/rusto/src/simple_adapter.rs
use crate::types::{IbddInAny, ITestResourceConfiguration, IUniversalTestAdapter, IArtifactory};
use async_trait::async_trait;

pub struct SimpleTestAdapter;

impl SimpleTestAdapter {
    pub fn new() -> Self {
        Self
    }
}

#[async_trait]
impl<I> IUniversalTestAdapter<I> for SimpleTestAdapter 
where
    I: IbddInAny + Send + Sync,
    <I as IbddInAny>::Iinput: Send + 'static,
    <I as IbddInAny>::Isubject: Send + 'static,
    <I as IbddInAny>::Istore: Send + 'static,
    <I as IbddInAny>::Given: Send + 'static,
    <I as IbddInAny>::Then: Send + 'static,
    <I as IbddInAny>::When: Send + 'static,
{
    async fn prepare_all(
        &self,
        input: I::Iinput,
        _tr: &ITestResourceConfiguration,
        _artifactory: Option<&IArtifactory>,
    ) -> I::Iinput {
        input
    }
    
    async fn prepare_each(
        &self,
        subject: I::Isubject,
        _initializer: I::Given,
        _test_resource: &ITestResourceConfiguration,
        _initial_values: &(dyn std::any::Any + Send + Sync),
        _artifactory: Option<&IArtifactory>,
    ) -> I::Isubject {
        subject
    }
    
    async fn cleanup_each(
        &self,
        store: I::Istore,
        _key: &str,
        _artifactory: Option<&IArtifactory>,
    ) -> I::Istore {
        store
    }
    
    async fn cleanup_all(
        &self,
        store: I::Istore,
        _artifactory: Option<&IArtifactory>,
    ) -> I::Istore {
        store
    }
    
    async fn execute(
        &self,
        store: I::Istore,
        _action_cb: I::When,
        _test_resource: &ITestResourceConfiguration,
        _artifactory: Option<&IArtifactory>,
    ) -> I::Istore {
        store
    }
    
    async fn verify(
        &self,
        store: I::Istore,
        _check_cb: I::Then,
        _test_resource: &ITestResourceConfiguration,
        _artifactory: Option<&IArtifactory>,
    ) -> I::Iselection {
        // For a simple adapter, we need to convert Istore to Iselection
        // This is a placeholder implementation
        // In practice, this would involve extracting a selection from the store
        // For now, we'll use a type conversion that may need to be adjusted
        // based on the actual types
        unsafe {
            std::mem::transmute_copy(&store)
        }
    }
    
    // Assertion - standardized name across all languages
    // Note: Named assert_that to avoid conflict with Java's 'assert' keyword
    fn assert_that(&self, _t: I::Then) -> bool {
        // Convert to bool - this is a simple implementation
        // In practice, this would depend on the actual type
        // For now, we'll assume it's a boolean or can be converted to one
        // This is a placeholder
        true
    }
}

File: src/lib/rubeno/lib/simple_adapter.rb
require 'types'

module Rubeno
  class SimpleTestAdapter
    include ITestAdapter
    
    # Universal adapter methods only
    def prepare_all(input_val, test_resource, artifactory = nil)
      puts "[SimpleTestAdapter] prepare_all called with artifactory: #{artifactory ? 'present' : 'nil'}"
      input_val
    end
    
    def prepare_each(subject, initializer, test_resource, initial_values, artifactory = nil)
      puts "[SimpleTestAdapter] prepare_each called with artifactory: #{artifactory ? 'present' : 'nil'}"
      if initializer.respond_to?(:call)
        initializer.call(subject)
      else
        subject
      end
    end
    
    def execute(store, action_cb, test_resource, artifactory = nil)
      puts "[SimpleTestAdapter] execute called with artifactory: #{artifactory ? 'present' : 'nil'}"
      if action_cb.respond_to?(:call)
        action_cb.call(store)
      else
        store
      end
    end
    
    def verify(store, check_cb, test_resource, artifactory = nil)
      puts "[SimpleTestAdapter] verify called with artifactory: #{artifactory ? 'present' : 'nil'}"
      if check_cb.respond_to?(:call)
        check_cb.call(store)
      else
        store
      end
    end
    
    def cleanup_each(store, key, artifactory = nil)
      puts "[SimpleTestAdapter] cleanup_each called with artifactory: #{artifactory ? 'present' : 'nil'}"
      store
    end
    
    def cleanup_all(store, artifactory = nil)
      puts "[SimpleTestAdapter] cleanup_all called with artifactory: #{artifactory ? 'present' : 'nil'}"
      store
    end
    
    # Assertion - standardized name across all languages
    def assert(x)
      !!x
    end
  end
end
